### PR TITLE
Adding statusio_maintenance module

### PR DESCRIPTION
**Pull request by bhcopeland**
Adding a new module to create maintenance for https://status.io/ dashboard

### CODE REVIEW

**Comment by robynbergeron** [comment]
Thanks @bhcopeland for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

**Comment by sivel** [comment]
Initial feedback, based on automated checks of this new module:

```
============================================================================
statusio_maintenance.py
============================================================================
Traceback (most recent call last):
  File "ansible_testing/modules.py", line 383, in validate
    doc, examples, ret = get_docstring(self.path, verbose=True)
  File "/Users/matt/python_venvs/ansibledev/ansible/lib/ansible/utils/module_docs.py", line 74, in get_docstring
    doc = yaml.safe_load(child.value.s)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/__init__.py", line 93, in safe_load
    return load(stream, SafeLoader)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/__init__.py", line 71, in load
    return loader.get_single_data()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/constructor.py", line 37, in get_single_data
    node = self.get_single_node()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 36, in get_single_node
    document = self.compose_document()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 55, in compose_document
    node = self.compose_node(None, None)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 82, in compose_node
    node = self.compose_sequence_node(anchor)
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/composer.py", line 110, in compose_sequence_node
    while not self.check_event(SequenceEndEvent):
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/parser.py", line 98, in check_event
    self.current_event = self.state()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/parser.py", line 382, in parse_block_sequence_entry
    if self.check_token(BlockEntryToken):
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/scanner.py", line 115, in check_token
    while self.need_more_tokens():
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/scanner.py", line 149, in need_more_tokens
    self.stale_possible_simple_keys()
  File "/Users/matt/python_venvs/ansibledev/lib/python2.7/site-packages/yaml/scanner.py", line 289, in stale_possible_simple_keys
    "could not found expected ':'", self.get_mark())
ScannerError: while scanning a simple key
  in "<string>", line 10, column 5:
        capture API traffic
        ^
could not found expected ':'
  in "<string>", line 11, column 5:
        - Use start_date and start_time  ...
        ^

ERROR: Invalid or no DOCUMENTATION provided
ERROR: No RETURN provided
ERROR: requests import found, should use ansible.module_utils.urls instead
```

**Comment by bhcopeland** [comment]
Thanks @sivel . Is above tests using https://github.com/sivel/ansible-testing ? Any documentation on using it, would quite fancy giving it a try.

Updated with new commit.

**Comment by sivel** [comment]
@bhcopeland it is.  I've just updated the README in that repo with more info.  Also see https://github.com/ansible/ansible-modules-extras/pull/1065

**Comment by bhcopeland** [comment]
Thanks @sivel. Nice Python script! Will keep that handy :+1:  Got it working and got most issues bar one fixed. It would be nice if it checked "examples" was in "documentation", as I had a typo in examples :) I spotted. I also noticed, when it notifies a indent error, the line number is wrong. Which is a bit of a issue when its on a popular field like "required: false". 

Maybe I am missing something, but I don't quite get what needs changing to fix "No RETURN provided".  

Also I will happily rebase and squash my commits for this PR.

**Comment by sivel** [comment]
@bhcopeland As far as "No RETURN provided", that is similar to `DOCUMENTATION` or `EXAMPLES` which you can read about at http://docs.ansible.com/ansible/developing_modules.html#example

`RETURN` is required to be provided with all new modules.

The line numbers are potentially off, I'm not sure I can fix that.  The line number is the line number within the YAML of `DOCUMENTATION`.

`EXAMPLES` isn't really parsed, it's just a free text area, so it would be hard to catch errors in `EXAMPLES`.

**Comment by bhcopeland** [comment]
@sivel Ah okay. That makes sense. I will add. Well done for getting your PR into Ansible/Travis.

**Comment by bhcopeland** [comment]
All done :+1:

**Comment by bhcopeland** [comment]
Is there anything else :)?

**Comment by bhcopeland** [comment]
Just updated some documentation errors and squashed my commits. 

ready_for_review

**Comment by gregdek** [comment]
Thanks @bhcopeland for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

**Comment by bhcopeland** [comment]
Hey Guys,

Can we get this pushed through, it'd really help.

Thanks

**Comment by bhcopeland** [comment]
bump :)?

**Comment by bhcopeland** [comment]
@bcoca, @gregdek @robynbergeron  would it be possible to make some headway on this?

File: monitoring/statusio_maintenance.py [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+    automation:
+        description:
+            - Automatically start and end the maintenance window
+        required: false
+        default: false
+    maintenance_notify_now:
+        description:
+            - Notify subscribers now
+        required: false
+        default: "https://api.status.io"
```
Comment (resmo) [inline monitoring/statusio_maintenance.py]: should be `default: false` IMHO

File: monitoring/statusio_maintenance.py [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+    maintenance_notify_72_hr:
+        description:
+            - Notify subscribers 72 hours before maintenance start time
+        required: false
+        default: "https://api.status.io"
```
Comment (resmo) [inline monitoring/statusio_maintenance.py]: should be `default: false` IMHO

File: monitoring/statusio_maintenance.py [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    maintenance_notify_24_hr:
+        description:
+            - Notify subscribers 24 hours before maintenance start time
+        required: false
+        default: "https://api.status.io"
```
Comment (resmo) [inline monitoring/statusio_maintenance.py]: should be `default: false` IMHO

File: monitoring/statusio_maintenance.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+    maintenance_notify_1_hr:
+        description:
+            - Notify subscribers 1 hour before maintenance start time
+        required: false
+        default: "https://api.status.io"
```
Comment (resmo) [inline monitoring/statusio_maintenance.py]: should be `default: false` IMHO

File: monitoring/statusio_maintenance.py [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+        description:
+            - The length of time in UTC that the maintenance will run \
+            (starting from playbook runtime)
+        required: false
+        default: 10
+    start_date:
+        description:
+            - Date maintenance is expected to start (Month/Day/Year) (UTC)
+            - End Date is worked out from start_date + minutes
+        required: false
```
Comment (resmo) [inline monitoring/statusio_maintenance.py]: add a default if `required: false`, e.g. `default: None`

File: monitoring/statusio_maintenance.py [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+    start_time:
+        description:
+            - Time maintenance is expected to start (Hour:Minutes) (UTC)
+            - End Time is worked out from start_time + minutes
+        required: false
```
Comment (resmo) [inline monitoring/statusio_maintenance.py]: add a default if `required: false`, e.g. `default: None`

File: monitoring/statusio_maintenance.py [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+            maintenance_notify_now=dict(type='bool', default=False,
+                                        required=False),
+            maintenance_notify_72_hr=dict(type='bool', default=False,
+                                          required=False),
+            maintenance_notify_24_hr=dict(type='bool', default=False,
+                                          required=False),
+            maintenance_notify_1_hr=dict(type='bool', default=False,
+                                         required=False),
+            maintenance_id=dict(required=False, default=None),
+            start_date=dict(default=False, required=False),
```
Comment (resmo) [inline monitoring/statusio_maintenance.py]: default should rather be None for start_date and start_time

**Comment by gregdek** [comment]
@bhcopeland A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

**Comment by bhcopeland** [comment]
Yeah I will make the changes asap. Haven't found time yet with  Easter weekend :)

**Comment by bhcopeland** [comment]
Changes have been made @gregdek. Let me know if there are any more, thanks @resmo.

File: monitoring/statusio_maintenance.py [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+     maintenance_id: "561f90faf74bc94a4700087b"
+     statuspage: "statuspage_id"
+     api_id: "api_id"
+     api_key: "api_key"
+     state: absent
+
+'''
+
+RETURN = '''
+changed:
```
Comment (resmo) [inline monitoring/statusio_maintenance.py]: remove changed as it is a common return

Comment (bhcopeland) [inline monitoring/statusio_maintenance.py]: removed, but now fails CI. I noticed some modules have RETURN whilst others don't (in the PR queue). Not sure what to put in the RETURN bit instead.

Comment (bhcopeland) [inline monitoring/statusio_maintenance.py]: @resmo any advice on this?

**Comment by gregdek** [comment]
@bhcopeland Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by bhcopeland** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @bhcopeland for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @bhcopeland for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by bhcopeland** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @bhcopeland for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by bhcopeland** [comment]
Bump :)

**Comment by resmo** [comment]
I am not able to test it but code looks solid.

shipit

**Comment by gregdek** [comment]
Thanks again to @bhcopeland for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by alikins** [comment]
merged to devel

**Pull request merged**